File: mini-services/trading-engine/app/smc/liquidity.py
"""
SMC Engine - Liquidity Detection Module
Mathematical implementation of Liquidity Zones, Equal Levels, Liquidity Sweeps

Mathematical Formulas:

1. Equal Highs Detection:
   if abs(high1 - high2) < threshold:
       equal_high = True
   threshold = 0.1% price difference

2. Liquidity Sweep Detection:
   if high > eq_high and close < eq_high:
       liquidity_sweep = True (Stop Hunt Detected)

3. Liquidity Strength:
   Based on number of touches at the level

Author: Trading AI Agent
"""

from dataclasses import dataclass
from typing import List, Optional, Tuple
from enum import Enum
from datetime import datetime
import numpy as np

from .swing import Swing, Candle


class LiquidityType(Enum):
    """Types of Liquidity Zones"""
    EQUAL_HIGHS = "EQUAL_HIGHS"
    EQUAL_LOWS = "EQUAL_LOWS"
    BUY_SIDE = "BUY_SIDE"      # Above swing highs (short SLs)
    SELL_SIDE = "SELL_SIDE"    # Below swing lows (long SLs)


@dataclass
class LiquidityZone:
    """Liquidity Zone"""
    type: LiquidityType
    price_level: float
    tolerance: float
    volume: float = 0.0
    swept: bool = False
    swept_at: Optional[datetime] = None
    touches: int = 0
    strength: float = 0.0


@dataclass
class LiquiditySweep:
    """Liquidity Sweep Result"""
    swept: bool
    zone: Optional[LiquidityZone]
    direction: str  # 'BULLISH' or 'BEARISH'
    timestamp: Optional[datetime] = None


class LiquidityDetector:
    """
    Liquidity Detection using mathematical formulas
    
    Smart Money hunts liquidity at:
    - Equal highs/lows (double tops/bottoms)
    - Above swing highs (buy-side)
    - Below swing lows (sell-side)
    """
    
    def __init__(self, equal_threshold: float = 0.1, volume_multiplier: float = 1.5):
        """
        Initialize Liquidity Detector
        
        Args:
            equal_threshold: Percentage threshold for equal levels (0.1%)
            volume_multiplier: Volume threshold for significant zones
        """
        self.equal_threshold = equal_threshold
        self.volume_multiplier = volume_multiplier
    
    def detect_equal_highs(self, swings: List[Swing], candles: List[Candle]) -> List[LiquidityZone]:
        """
        Detect Equal Highs (Double/Triple Tops)
        
        Mathematical Formula:
        if abs(high1 - high2) / avg_price * 100 < threshold:
            equal_high = True
        
        Creates liquidity pool where stop losses cluster
        """
        swing_highs = [s for s in swings if s.type == 'HIGH']
        
        if len(swing_highs) < 2:
            return []
        
        zones = []
        used_indices = set()
        
        for i in range(len(swing_highs)):
            if i in used_indices:
                continue
            
            for j in range(i + 1, len(swing_highs)):
                if j in used_indices:
                    continue
                
                swing1 = swing_highs[i]
                swing2 = swing_highs[j]
                
                # Mathematical condition for equal highs
                price_diff = abs(swing1.price - swing2.price)
                avg_price = (swing1.price + swing2.price) / 2
                percent_diff = (price_diff / avg_price) * 100
                
                if percent_diff <= self.equal_threshold:
                    # Check if swept
                    swept, swept_at = self._check_sweep(candles, avg_price, swing2, 'HIGH')
                    
                    # Count touches
                    touches = self._count_touches(candles, avg_price, 'HIGH')
                    
                    zone = LiquidityZone(
                        type=LiquidityType.EQUAL_HIGHS,
                        price_level=avg_price,
                        tolerance=self.equal_threshold,
                        swept=swept,
                        swept_at=swept_at,
                        touches=touches,
                        strength=self._calculate_strength(touches)
                    )
                    zones.append(zone)
                    used_indices.add(i)
                    used_indices.add(j)
                    break
        
        return zones
    
    def detect_equal_lows(self, swings: List[Swing], candles: List[Candle]) -> List[LiquidityZone]:
        """
        Detect Equal Lows (Double/Triple Bottoms)
        
        Mathematical Formula:
        if abs(low1 - low2) / avg_price * 100 < threshold:
            equal_low = True
        """
        swing_lows = [s for s in swings if s.type == 'LOW']
        
        if len(swing_lows) < 2:
            return []
        
        zones = []
        used_indices = set()
        
        for i in range(len(swing_lows)):
            if i in used_indices:
                continue
            
            for j in range(i + 1, len(swing_lows)):
                if j in used_indices:
                    continue
                
                swing1 = swing_lows[i]
                swing2 = swing_lows[j]
                
                price_diff = abs(swing1.price - swing2.price)
                avg_price = (swing1.price + swing2.price) / 2
                percent_diff = (price_diff / avg_price) * 100
                
                if percent_diff <= self.equal_threshold:
                    swept, swept_at = self._check_sweep(candles, avg_price, swing2, 'LOW')
                    touches = self._count_touches(candles, avg_price, 'LOW')
                    
                    zone = LiquidityZone(
                        type=LiquidityType.EQUAL_LOWS,
                        price_level=avg_price,
                        tolerance=self.equal_threshold,
                        swept=swept,
                        swept_at=swept_at,
                        touches=touches,
                        strength=self._calculate_strength(touches)
                    )
                    zones.append(zone)
                    used_indices.add(i)
                    used_indices.add(j)
                    break
        
        return zones
    
    def detect_buy_side_liquidity(self, swings: List[Swing], 
                                  candles: List[Candle]) -> List[LiquidityZone]:
        """
        Detect Buy-Side Liquidity (Above swing highs)
        
        Stop losses for short positions sit above swing highs
        Smart money hunts these levels
        """
        swing_highs = sorted([s for s in swings if s.type == 'HIGH'],
                           key=lambda s: s.candle_index)
        
        if not swing_highs:
            return []
        
        zones = []
        
        # Get recent swing highs (last 5)
        recent_highs = swing_highs[-5:] if len(swing_highs) >= 5 else swing_highs
        
        for swing in recent_highs:
            swept, swept_at = self._check_sweep(candles, swing.price, swing, 'HIGH')
            
            zone = LiquidityZone(
                type=LiquidityType.BUY_SIDE,
                price_level=swing.price,
                tolerance=self.equal_threshold,
                swept=swept,
                swept_at=swept_at,
                strength=swing.strength
            )
            zones.append(zone)
        
        return zones
    
    def detect_sell_side_liquidity(self, swings: List[Swing],
                                   candles: List[Candle]) -> List[LiquidityZone]:
        """
        Detect Sell-Side Liquidity (Below swing lows)
        
        Stop losses for long positions sit below swing lows
        """
        swing_lows = sorted([s for s in swings if s.type == 'LOW'],
                          key=lambda s: s.candle_index)
        
        if not swing_lows:
            return []
        
        zones = []
        recent_lows = swing_lows[-5:] if len(swing_lows) >= 5 else swing_lows
        
        for swing in recent_lows:
            swept, swept_at = self._check_sweep(candles, swing.price, swing, 'LOW')
            
            zone = LiquidityZone(
                type=LiquidityType.SELL_SIDE,
                price_level=swing.price,
                tolerance=self.equal_threshold,
                swept=swept,
                swept_at=swept_at,
                strength=swing.strength
            )
            zones.append(zone)
        
        return zones
    
    def detect_all_liquidity(self, swings: List[Swing], 
                            candles: List[Candle]) -> List[LiquidityZone]:
        """Detect all liquidity zones"""
        equal_highs = self.detect_equal_highs(swings, candles)
        equal_lows = self.detect_equal_lows(swings, candles)
        buy_side = self.detect_buy_side_liquidity(swings, candles)
        sell_side = self.detect_sell_side_liquidity(swings, candles)
        
        return equal_highs + equal_lows + buy_side + sell_side
    
    def detect_liquidity_sweep(self, candles: List[Candle], 
                               zones: List[LiquidityZone],
                               lookback: int = 5) -> LiquiditySweep:
        """
        Detect Liquidity Sweep (Stop Hunt)
        
        Mathematical Condition:
        - Bullish Sweep: high > level and close < level
        - Bearish Sweep: low < level and close > level
        
        This indicates smart money swept stops before reversing
        """
        if len(candles) < lookback or not zones:
            return LiquiditySweep(swept=False, zone=None, direction='NONE')
        
        recent_candles = candles[-lookback:]
        unswept_zones = [z for z in zones if not z.swept]
        
        for candle in recent_candles:
            for zone in unswept_zones:
                if zone.type in [LiquidityType.EQUAL_HIGHS, LiquidityType.BUY_SIDE]:
                    # Bullish sweep: price breaks above and closes below
                    if candle.high > zone.price_level and candle.close < zone.price_level:
                        return LiquiditySweep(
                            swept=True,
                            zone=zone,
                            direction='BULLISH',
                            timestamp=candle.timestamp
                        )
                else:
                    # Bearish sweep: price breaks below and closes above
                    if candle.low < zone.price_level and candle.close > zone.price_level:
                        return LiquiditySweep(
                            swept=True,
                            zone=zone,
                            direction='BEARISH',
                            timestamp=candle.timestamp
                        )
        
        return LiquiditySweep(swept=False, zone=None, direction='NONE')
    
    def _check_sweep(self, candles: List[Candle], price_level: float,
                    after_swing: Swing, direction: str) -> Tuple[bool, Optional[datetime]]:
        """Check if a liquidity zone has been swept"""
        for candle in candles[after_swing.candle_index + 1:]:
            if direction == 'HIGH':
                if candle.high > price_level and candle.close < price_level:
                    return True, candle.timestamp
            else:
                if candle.low < price_level and candle.close > price_level:
                    return True, candle.timestamp
        
        return False, None
    
    def _count_touches(self, candles: List[Candle], price_level: float, 
                       direction: str) -> int:
        """Count touches at a price level"""
        tolerance = price_level * (self.equal_threshold / 100)
        touches = 0
        
        for candle in candles:
            if direction == 'HIGH':
                if abs(candle.high - price_level) <= tolerance:
                    touches += 1
            else:
                if abs(candle.low - price_level) <= tolerance:
                    touches += 1
        
        return touches
    
    def _calculate_strength(self, touches: int) -> float:
        """Calculate liquidity strength based on touches"""
        # Normalize to 0-1 scale, max at 5 touches
        return min(touches / 5, 1.0)


def get_nearest_liquidity(price: float, zones: List[LiquidityZone],
                         zone_type: Optional[LiquidityType] = None) -> Optional[LiquidityZone]:
    """Get nearest liquidity zone to current price"""
    filtered = zones
    if zone_type:
        filtered = [z for z in zones if z.type == zone_type]
    
    if not filtered:
        return None
    
    # Sort by distance from price
    return min(filtered, key=lambda z: abs(z.price_level - price))

File: src/lib/trading/smc/liquidity.ts
// SMC Engine - Liquidity Detection Module
// Detects liquidity pools, equal highs/lows, and liquidity sweeps

import { Candle, LiquidityZone, LiquidityType, Swing, SwingType } from '../types';

export interface LiquidityConfig {
  equalThreshold: number;  // Percentage threshold for equal highs/lows
  volumeMultiplier: number; // Volume threshold for significant zones
  lookbackPeriod: number;  // Number of candles to look back
}

const DEFAULT_CONFIG: LiquidityConfig = {
  equalThreshold: 0.1,  // 0.1% difference
  volumeMultiplier: 1.5,
  lookbackPeriod: 50
};

/**
 * Detect Equal Highs (Double/Triple Tops)
 * Equal highs create liquidity pools where stop losses cluster
 */
export function detectEqualHighs(
  candles: Candle[],
  swings: Swing[],
  config: LiquidityConfig = DEFAULT_CONFIG
): LiquidityZone[] {
  const zones: LiquidityZone[] = [];
  const swingHighs = swings.filter(s => s.type === SwingType.HIGH);

  if (swingHighs.length < 2) return zones;

  for (let i = 0; i < swingHighs.length - 1; i++) {
    for (let j = i + 1; j < swingHighs.length; j++) {
      const swing1 = swingHighs[i];
      const swing2 = swingHighs[j];

      const priceDiff = Math.abs(swing1.price - swing2.price);
      const avgPrice = (swing1.price + swing2.price) / 2;
      const percentDiff = (priceDiff / avgPrice) * 100;

      if (percentDiff <= config.equalThreshold) {
        // Check if already swept
        const swept = isZoneSwept(candles, avgPrice, swing2.timestamp, 'HIGH');

        zones.push({
          symbolId: swing1.symbolId,
          timeframe: swing1.timeframe,
          type: LiquidityType.EQUAL_HIGHS,
          priceLevel: avgPrice,
          tolerance: config.equalThreshold,
          volume: getVolumeAtLevel(candles, avgPrice),
          swept: swept,
          sweptAt: swept ? findSweepTime(candles, avgPrice, swing2.timestamp, 'HIGH') : undefined
        });
      }
    }
  }

  return zones;
}

/**
 * Detect Equal Lows (Double/Triple Bottoms)
 */
export function detectEqualLows(
  candles: Candle[],
  swings: Swing[],
  config: LiquidityConfig = DEFAULT_CONFIG
): LiquidityZone[] {
  const zones: LiquidityZone[] = [];
  const swingLows = swings.filter(s => s.type === SwingType.LOW);

  if (swingLows.length < 2) return zones;

  for (let i = 0; i < swingLows.length - 1; i++) {
    for (let j = i + 1; j < swingLows.length; j++) {
      const swing1 = swingLows[i];
      const swing2 = swingLows[j];

      const priceDiff = Math.abs(swing1.price - swing2.price);
      const avgPrice = (swing1.price + swing2.price) / 2;
      const percentDiff = (priceDiff / avgPrice) * 100;

      if (percentDiff <= config.equalThreshold) {
        const swept = isZoneSwept(candles, avgPrice, swing2.timestamp, 'LOW');

        zones.push({
          symbolId: swing1.symbolId,
          timeframe: swing1.timeframe,
          type: LiquidityType.EQUAL_LOWS,
          priceLevel: avgPrice,
          tolerance: config.equalThreshold,
          volume: getVolumeAtLevel(candles, avgPrice),
          swept: swept,
          sweptAt: swept ? findSweepTime(candles, avgPrice, swing2.timestamp, 'LOW') : undefined
        });
      }
    }
  }

  return zones;
}

/**
 * Detect Buy-Side Liquidity (Above swing highs)
 * Stop losses for short positions sit above swing highs
 */
export function detectBuySideLiquidity(
  candles: Candle[],
  swings: Swing[],
  config: LiquidityConfig = DEFAULT_CONFIG
): LiquidityZone[] {
  const zones: LiquidityZone[] = [];
  const swingHighs = swings.filter(s => s.type === SwingType.HIGH);

  if (swingHighs.length === 0) return zones;

  // Find recent swing highs that haven't been swept
  const recentHighs = swingHighs.slice(-5);

  for (const swing of recentHighs) {
    const swept = isZoneSwept(candles, swing.price, swing.timestamp, 'HIGH');

    zones.push({
      symbolId: swing.symbolId,
      timeframe: swing.timeframe,
      type: LiquidityType.BUY_SIDE,
      priceLevel: swing.price,
      tolerance: config.equalThreshold,
      volume: swing.strength,
      swept: swept,
      sweptAt: swept ? findSweepTime(candles, swing.price, swing.timestamp, 'HIGH') : undefined
    });
  }

  return zones;
}

/**
 * Detect Sell-Side Liquidity (Below swing lows)
 * Stop losses for long positions sit below swing lows
 */
export function detectSellSideLiquidity(
  candles: Candle[],
  swings: Swing[],
  config: LiquidityConfig = DEFAULT_CONFIG
): LiquidityZone[] {
  const zones: LiquidityZone[] = [];
  const swingLows = swings.filter(s => s.type === SwingType.LOW);

  if (swingLows.length === 0) return zones;

  const recentLows = swingLows.slice(-5);

  for (const swing of recentLows) {
    const swept = isZoneSwept(candles, swing.price, swing.timestamp, 'LOW');

    zones.push({
      symbolId: swing.symbolId,
      timeframe: swing.timeframe,
      type: LiquidityType.SELL_SIDE,
      priceLevel: swing.price,
      tolerance: config.equalThreshold,
      volume: swing.strength,
      swept: swept,
      sweptAt: swept ? findSweepTime(candles, swing.price, swing.timestamp, 'LOW') : undefined
    });
  }

  return zones;
}

/**
 * Detect all liquidity zones
 */
export function detectLiquidity(
  candles: Candle[],
  swings: Swing[],
  config: LiquidityConfig = DEFAULT_CONFIG
): LiquidityZone[] {
  const equalHighs = detectEqualHighs(candles, swings, config);
  const equalLows = detectEqualLows(candles, swings, config);
  const buySide = detectBuySideLiquidity(candles, swings, config);
  const sellSide = detectSellSideLiquidity(candles, swings, config);

  return [...equalHighs, ...equalLows, ...buySide, ...sellSide];
}

/**
 * Check if a liquidity zone has been swept
 */
function isZoneSwept(
  candles: Candle[],
  priceLevel: number,
  afterTime: Date,
  direction: 'HIGH' | 'LOW'
): boolean {
  const afterCandles = candles.filter(c => c.timestamp > afterTime);

  for (const candle of afterCandles) {
    if (direction === 'HIGH') {
      // Swept if price went above and closed below
      if (candle.high > priceLevel && candle.close < priceLevel) {
        return true;
      }
    } else {
      // Swept if price went below and closed above
      if (candle.low < priceLevel && candle.close > priceLevel) {
        return true;
      }
    }
  }

  return false;
}

/**
 * Find the time when sweep occurred
 */
function findSweepTime(
  candles: Candle[],
  priceLevel: number,
  afterTime: Date,
  direction: 'HIGH' | 'LOW'
): Date | undefined {
  const afterCandles = candles.filter(c => c.timestamp > afterTime);

  for (const candle of afterCandles) {
    if (direction === 'HIGH') {
      if (candle.high > priceLevel && candle.close < priceLevel) {
        return candle.timestamp;
      }
    } else {
      if (candle.low < priceLevel && candle.close > priceLevel) {
        return candle.timestamp;
      }
    }
  }

  return undefined;
}

/**
 * Get volume at a price level
 */
function getVolumeAtLevel(candles: Candle[], priceLevel: number): number {
  let totalVolume = 0;
  const tolerance = priceLevel * 0.001; // 0.1% tolerance

  for (const candle of candles) {
    if (candle.high >= priceLevel - tolerance && candle.low <= priceLevel + tolerance) {
      totalVolume += candle.volume;
    }
  }

  return totalVolume;
}

/**
 * Detect liquidity sweep pattern (stop hunt)
 * Returns true if a sweep has just occurred
 */
export function detectLiquiditySweep(
  candles: Candle[],
  zones: LiquidityZone[],
  lookback: number = 5
): { swept: boolean; zone: LiquidityZone | null; direction: 'BULLISH' | 'BEARISH' | null } {
  if (candles.length < lookback || zones.length === 0) {
    return { swept: false, zone: null, direction: null };
  }

  const recentCandles = candles.slice(-lookback);
  const unsweptZones = zones.filter(z => !z.swept);

  for (const zone of unsweptZones) {
    for (const candle of recentCandles) {
      if (zone.type === LiquidityType.EQUAL_HIGHS || zone.type === LiquidityType.BUY_SIDE) {
        // Bullish sweep: price breaks above and closes below
        if (candle.high > zone.priceLevel && candle.close < zone.priceLevel) {
          return { swept: true, zone, direction: 'BULLISH' };
        }
      } else {
        // Bearish sweep: price breaks below and closes above
        if (candle.low < zone.priceLevel && candle.close > zone.priceLevel) {
          return { swept: true, zone, direction: 'BEARISH' };
        }
      }
    }
  }

  return { swept: false, zone: null, direction: null };
}

/**
 * Calculate liquidity strength based on touches
 */
export function calculateLiquidityStrength(
  candles: Candle[],
  zone: LiquidityZone
): number {
  let touches = 0;
  const tolerance = zone.priceLevel * (zone.tolerance / 100);

  for (const candle of candles) {
    const priceRange = Math.abs(candle.high - candle.low);
    
    if (zone.type === LiquidityType.EQUAL_HIGHS || zone.type === LiquidityType.BUY_SIDE) {
      if (Math.abs(candle.high - zone.priceLevel) <= tolerance) {
        touches++;
      }
    } else {
      if (Math.abs(candle.low - zone.priceLevel) <= tolerance) {
        touches++;
      }
    }
  }

  return Math.min(touches / 3, 1); // Normalize to 0-1
}
